RENEWED: Extending fluid concept to fit CF & Standard formulations

vlipovac opened the pull request:
Closes #1240 
Closes #1247

Replaces #1244 for reasons of clarity.

## Proposed changes

Reworking the concept of `model.fluid` to be a general fluid with phases and components, instead of just a data collection of constants.

Breaking change with a multiple minor changes throughout the packages. The essential changes are summarized in a self-review in #1244.

`porepy.compositional.ChemicalSpecies` was merged into `FluidConstants`, including data class functionality. This led to some changes for `SolidConstants` as well for inheritance reasons.

The default model set-up is changed to provide a 1-phase, 1-component fluid, where the single component is based on the user-provided `FluidConstants` in the `model.params`.
The interface to various fluid properties (e.g., `fluid_density`) is replaced by the general `Fluid` class.

## Types of changes

- [ ] Minor change (e.g., dependency bumps, broken links).
- [ ] Bugfix (non-breaking change which fixes an issue).
- [ ] New feature (non-breaking change which adds functionality).
- [X] Breaking change (fix or feature that would cause existing functionality to not work as expected).
- [X] Testing (contribution related to testing of existing or new functionality).
- [ ] Documentation (contribution related to adding, improving, or fixing documentation).
- [X] Maintenance (e.g., improve logic and performance, remove obsolete code).
- [ ] Other:

## Checklist

- [X] The documentation is up-to-date.
- [X] Static typing is included in the update.
- [X] This PR does not duplicate existing functionality.
- [X] The update is covered by the test suite (including tests added in the PR).
- [X] If new skipped tests have been introduced in this PR, `pytest` was run with the `--run-skipped` flag.

IvarStefansson commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1022,66 +1021,64 @@ def fraction(domains: pp.SubdomainsOrBoundaries) -> pp.ad.Operator:
 
-    Provides means to create domain-dependent callables representing thermodynamic
-    properties which appear in equations.
+    The following methods provide expressions for them:
```
What does "them" refer to?

vlipovac replied:
Phase properties. Will make this more clear.

vlipovac replied:
Done.

IvarStefansson commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1095,25 +1092,39 @@ def create_mixture(self) -> None:
+
+        """
+        # Should be available after SolutionStrategy.set_materials()
+        # Getting the user-passed fluid constants to create the default fluid component
```
This is confusing. Why are the user-passed constants private?

vlipovac replied:
The user-passed constants are now used to create the 1 component for the default case of a 1-phase, 1-component fluid. I wanted to hint to developers to not mess with this objects, which I stored in `params` until the `model.fluid` is created.

Related to your previous comment https://github.com/pmgbergen/porepy/pull/1252#discussion_r1822352597

I can add more documentation to explain this, but it belongs to the inner workings and is (hopefully) only exposed to developers digging pretty deep. I am also open for a better mechanism to transfer the fluid constants from the `params` dicts provided in the instantiation of the solution strategy, to this point in the chain `prepare_simulation() -> set_materials() -> create_fluid()`

vlipovac replied:
Or maybe I should just use `params['material_constants']['fluid']` ...

vlipovac replied:
I removed the ominous `_default_fluid_constans` and rely solely on `self.params['material_constants']['fluid']`, also in the FluidMixin class.

IvarStefansson commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1217,102 @@ def dependencies_of_phase_properties(
-            ]
-        else:
+        Important:
+            This method must be overwritten in every flow problem which relies on e.g.,
```
It is not clear to me what the surrogate factory is an example of. I.e., I don't know in which other cases this comment is important.

vlipovac replied:
In cases with more than 1 phase or component, which relies on external flash calculations or look-ups like the OBL approach. Will it help to add these examples?

It boils down to the user having to provide information to the framework, on which variables the fluid properties depend
(`rho(p,T,z), rho(p,h,z)` etc.) This information is used to evaluate those variables and feed them to the flash or the look-up.

keileg replied:
Please provide the gist of this explanation to the dostring.

vlipovac replied:
See `Example:` section in the docstring.
This will be covered by the tutorial from #1254, where we need to revisit this docstring and shorten it respectively.

keileg replied:
I modified the documentation slightly to make it fit with my understanding. @vlipovac Could you verify that what I wrote is correct, and if so resolve this discussion?

IvarStefansson commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1217,102 @@ def dependencies_of_phase_properties(
-                + " equilibrium conditions defined. Per default, phase fractions are"
-                + " only created when an equilibrium formulation is used in the model."
-            )
+                class MyMixin:
```
I think we should aim at migrating some of these examples to a tutorial

vlipovac replied:
Yes, this will be part of the coming tutorial on tracer flow, which will be part of #1236

IvarStefansson commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -1222,106 +1217,102 @@ def dependencies_of_phase_properties(
-        self, phase: Phase
-    ) -> Callable[[pp.SubdomainsOrBoundaries], pp.ad.Operator]:
+    def specific_volume_of_phase(self, phase: Phase) -> ExtendedDomainFunctionType:
         """The specific volume of the phase is returned as a function calling the
```
Please add parameters and return here and below.

vlipovac replied:
Parameters and return values added.

IvarStefansson commented on src/porepy/compositional/base.py:
```diff
@@ -153,11 +161,11 @@ class Compound(Component):
     the :attr:`~Component.fraction` of the compound, i.e. the moles/mass of them are
     given by a product of mixture density, compound fraction and tracer fraction.
     :attr:`tracer_fraction_of` are assigned by the AD interface
```
What does "simplified, but meaningfully generalized," mean (start of docstring)?

vlipovac replied:
Many fluid models out there like brine models or the black-oil model, treat multiple chemical species which are in reality all separate species, bundled into an artificial component (`Compound`).
For example H2O-NaCl brine is often treated as a such a bundled component, thought strictly speaking one should have H2O and two ion species. But the simplifications suffices for some target applications.

keileg replied:
Please add the gist of this explanation to the docstring.

keileg replied:
Thanks. I am fine with resolving.

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,476 @@
+    Every derived data class must have one implemented and annotated as a ``ClassVar``.
+
+    Note:
+        The ``ClassVar`` hints to ``dataclass`` that this field should not be part of
```
The class attribute declares the SI units, e.g., for pressure `Pa`, contrary to the (non-SI) simulation units, they are universal in some sense.
The motivation was to have that declared in 1 spot in the code (programmatically accessible), instead of writing it in the docs *and* hardcoding it in the previously callable attributes.

Yes, it prevents you to have, say, two fluid constants where one has `Pa` and the other `MPa` or something else declared as the base unit.

Also, the unit conversion is done in a general manner for all attributes, from here declared SI units to simulation units.

I also wanted to make it more straight forward to add new constants to a class, without having to worry about conversion etc. Further development now needs essentially to declare simply the field, and add its unit to this attribute (like I did here for the critical pressure for example).

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,476 @@
+            The default value of ``units`` allows a conversion to SI units.
+            When using units which are not defined in the standard
+            :class:`~porepy.models.units.Units` class, the default value of ``units``
+            might not be enough.
```
The user must be aware which `Units` object was passed at instantiation.
If it was one which declared additional units to the ones defined in the base `Units` class, the conversion to SI (with the default value of `units`) will fail because it is the base class.

So if you have some `MyUnits` class, which was passed like `f = FluidConstants(units=MyUnits(...))`, then you also need `f.to_units(MyUnits())` to convert `f` back to SI (if you had non-SI when instantiating `f`)

Also, this default value here matches the default value of the `f.units` field.

I am open to not providing a default value here, to force the user to remember whether custom units classes were used or not.

IvarStefansson replied:
Perhaps removing the default is a good idea. Cause it's not going to hurt much (you can always pass pp.Units()). I also am a bit afraid of having a default which is an instance, which, I believe, will only be instantiated once when importing (right?). Also, do you think this note might just as well go into the parameter description, since it relates strictly to the parameter ``units``?

vlipovac replied:
Removed the default value.

vlipovac commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,476 @@
+
+    Parameters:
+        names: A list of names or chemical formulae to look up the chemical species.
+        package: ``default='chemicals'``
```
This function was purely motivated by my own use, and I only considered `chemicals` because it is the most complete package. We can remove the default value though, since `chemicals` is not a pp dependency.

vlipovac replied:
Removed the default value.

vlipovac commented on src/porepy/applications/md_grids/model_geometries.py:
```diff
@@ -7,7 +7,7 @@
 
 
-class SquareDomainOrthogonalFractures:
+class SquareDomainOrthogonalFractures(pp.ModelGeometry):
```
I saw that it was done at multiple other occasions where special grids were constructed, I guess for the same reasons I did it here: to have remove the declaration of mixed-in methods by help of protocol.

IvarStefansson replied:
Yes, I noticed that as well as I continued my review. Would it make sense to replace ModelGeometry by the protocol?

vlipovac replied:
Yes, let's do that in a separate issue, like you suggested here https://github.com/pmgbergen/porepy/pull/1252#discussion_r1824164946

vlipovac commented on src/porepy/models/fluid_mass_balance.py:
```diff
@@ -41,8 +41,10 @@ class MassBalanceEquations(pp.BalanceEquation):
     """
-    fluid_density: Callable[[pp.SubdomainsOrBoundaries], pp.ad.Operator]
-    """Fluid density. Defined in a mixin class with a suitable constitutive relation.
+    equation_system: pp.ad.EquationSystem
```
Indeed,  can be removed.

vlipovac replied:
Removed the redundant declaration.

vlipovac commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -2261,26 +2010,11 @@ def thermal_conductivity(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
             size = sum(sd.num_cells for sd in subdomains)
             phi = phi * pp.wrap_as_dense_ad_array(1, size)
-        conductivity = phi * self.fluid_thermal_conductivity(subdomains) + (
+        conductivity = phi * self.fluid.thermal_conductivity(subdomains) + (
```
I think so, until the point where the solid is also generalized, which I think is inevitable (keywords reactions, dissolutions, crystalization)

vlipovac commented on src/porepy/models/fluid_mass_balance.py:
```diff
@@ -627,7 +631,7 @@ def reference_pressure(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
 
         """
-        p_ref = self.fluid.pressure()
+        p_ref = self.fluid.reference_component.pressure
```
Related to https://github.com/pmgbergen/porepy/pull/1252#discussion_r1823359429 and potential further work.

vlipovac replied:
Resolve?

keileg replied:
Should we make an issue first? This is possibly related to the other unresolved comment.

vlipovac replied:
Is there another issue besides the one we fixed by introducing `ReferenceVariableValues`?

vlipovac commented on src/porepy/models/fluid_mass_balance.py:
```diff
@@ -824,6 +828,7 @@ class SinglePhaseFlow(  # type: ignore[misc]
     ConstitutiveLawsSinglePhaseFlow,
     BoundaryConditionsSinglePhaseFlow,
     SolutionStrategySinglePhaseFlow,
+    pp.FluidMixin,
```
The creation of the objects in the fluid per se (components, phases etc) is completely orthogonal to the rest, so it doesn't matter.

vlipovac commented on src/porepy/models/momentum_balance.py:
```diff
@@ -865,6 +865,7 @@ class MomentumBalance(  # type: ignore[misc]
     ConstitutiveLawsMomentumBalance,
     BoundaryConditionsMomentumBalance,
     SolutionStrategyMomentumBalance,
+    pp.FluidMixin,
```
The mechanics-only model is a unique in the sense, that it doesn't need as you said.
But since the respective solution strategy inherits from the base `SolutionStrategy`, the base class would require special logic to treat this unique case.

I then realized, that the mechanics-only model so far also had a redundant `fluid: FluidConstants` (probably for the same reasons). So I decided to go with that and now it has a redundant `fluid: Fluid` attribute. But since that one requires some this mixin, this redundancy became more visible now.

IvarStefansson replied:
I see. What part of the SolutionStrategy requires the fluid, specifically?

vlipovac replied:
The `prepare_simulation() -> set_materials()` call-stack, where `solid` and `fluid` are created.
Here we would need some, as I see it, ugly code checking base classes and inheritance, if we want a conditional creation of the fluid.
Ugly, because I cannot foresee how many multi-physics model will come in addition to the existing ones, and how to programmatically determine that only `MomentumBalance` is used, without altering the code every time a new model is introduced inheriting from `MomentumBalance`.

IvarStefansson replied:
I see. I can cook up an alternative solution. But I don't think it has any significant benefits, so let's leave as is.

keileg replied:
I added a short comment in the code.

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -48,19 +48,30 @@ def __init__(self, params: Optional[dict] = None):
 
         default_params.update(params)
         self.params = default_params
+        """Dictionary of parameters."""
```
@keileg  noticed that some docstrings were purged by accident in the protocol PR. I brought them back.
@Yuriyzabegaev  mentioned that he will explain in the next user meeting that we aim for double documentation?

Yuriyzabegaev replied:
Thanks! Yes, they should be duplicated.

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -112,6 +124,12 @@ def prepare_simulation(self) -> None:
         self.create_variables()
+        # After fluid and variables are defined, we can define the secondary quantities
+        # like fluid properties (which depend on variables).
+        # NOTE this is critical in the case where properties depend on some fractions.
```
The order of methods executed in `prepare_simulation`, yes.
Like you can't call `create_variables` before `set_equation_system_manager`, you can't define secondary terms like density without primary terms like variables.

IvarStefansson replied:
Thanks. I updated the explanation for clarity.

vlipovac commented on src/porepy/models/units.py:
```diff
@@ -123,3 +125,73 @@ def W(self) -> number:
+        self, value: number, units: str, to_si: Optional[bool] = False
+    ) -> number: ...
+
+    @overload
```
That code is not from me, I just moved it from `MaterialConstants` to `Units` according to #1247 .
Could you specify what is missing?

IvarStefansson replied:
Damn! I might be the culprit myself. Isn't the purpose for mypy to know that if the method is called with value being a number, it will return a number? And same for ndarray. By the way, I think the same could be more elegantly achieved using a typevar or something of the sort.

vlipovac replied:
Yes, I mean it's working! And yes that is possible, with much less code.
`T = TypeVar('T', pp.number, np.ndarray)` would do the trick.

vlipovac replied:
Removed the overload and introduced a local type var, to be either a number or an array.

vlipovac commented on src/porepy/utils/porepy_types.py:
```diff
@@ -35,3 +37,23 @@
+
+"""
+
+ExtendedDomainFunctionType = Union[DomainFunctionType, "pp.ad.SurrogateFactory"]
```
Can't think unfortunately of anything without turning this into a monster of a name.

vlipovac commented on tests/models/test_constitutive_laws.py:
```diff
@@ -365,7 +371,14 @@ def test_evaluated_values(
+    for name in method_namespace:
+        method = getattr(owner, name)
+        owner = method
+    # method = getattr(setup, method_name)
```
Because of the change from `model.fluid*_*density` to `model.fluid*.*density`: There is 1 more level in the name space.
First owner: `model`, second owner: `fluid`, and only then you come to the method which requires testing according to *tested vs testable methods*

vlipovac commented on src/porepy/examples/flow_benchmark_2d_case_1.py:
```diff
@@ -12,43 +12,36 @@
 
-class FractureSolidConstants(pp.SolidConstants):
+@dataclass(frozen=True, kw_only=True)
+class FractureSolidConstants(SolidConstants):
```
I think there is no part in the tutorial explicitly for material constants.
Since the general `fluid` itself needs one, could we put it there? This is probably a new user group issue.

IvarStefansson replied:
Sounds good. Could you create and label the issue, please?

vlipovac replied:
Done!

vlipovac commented on src/porepy/examples/flow_benchmark_3d_case_3.py:
```diff
@@ -87,27 +85,14 @@ def intersection_permeability(self, subdomains: list[pp.Grid]) -> pp.ad.Operator
 
 
-class BoundaryConditions:
+class BoundaryConditions(pp.BoundaryConditionMixin):
```
Yes I think it would achieve the same.
This is related to https://github.com/pmgbergen/porepy/pull/1252#discussion_r1822219015
Should we make the general change for all the available mixins, providing examples for BC, grids, etc.?

IvarStefansson replied:
Yes, I support that. Should we do it as a separate PR, though?

vlipovac replied:
Yes, I think another issue should document this.

keileg replied:
Has the issue been made? When so, feel free to resolve.

vlipovac replied:
Created #1260

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -26,21 +26,43 @@
+
+class _MixtureDOFHandler(PorePyModel):
     """A class to help resolve the independent fractional variables of an arbitrary
     mixture, and respectivly the DOFs.
```
If you model a fluid more complex than 1 phase and 1 component, you automatically get some unknowns into the model, related to the number of phases or components, which is then not 1.

keileg replied:
Please add the gist of this comment to the docstring.

vlipovac replied:
Improved class docstring.

IvarStefansson commented on src/porepy/models/energy_balance.py:
```diff
@@ -271,6 +265,23 @@ def interface_energy_flux(self, interfaces: list[pp.MortarGrid]) -> pp.ad.Operat
         flux.set_name("interface_energy_flux")
         return flux
 
+    def mobility_rho_h(self, domains: pp.SubdomainsOrBoundaries) -> pp.ad.Operator:
```
Should we make a note in 1236? Anyway, feel free to resolve

vlipovac replied:
Noted in draft #1236 .

vlipovac commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,337 @@
+""""Module containing some constant heuristic fluid property implemenations.
+
+Most of the laws implemented here are meant for 1-phase, 1-component mixtures, using some
+fluid constants stored in the fluid's reference component.
```
Information added to docstrings of affected classes.
There are some classes which work for single-phase and are agnostic to multiple components. I narrowed down the single-phase and or single-component case respectively in each class.

vlipovac commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,337 @@
+    """Fluid density as a function of temperature."""
+
+    def fluid_thermal_expansion(self, subdomains: Sequence[pp.Grid]) -> pp.ad.Operator:
+        """
```
Done.

vlipovac commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,337 @@
+    """
+
+    def fluid_specific_heat_capacity(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
+        """
```
Done.

vlipovac commented on src/porepy/models/protocol.py:
```diff
@@ -2,6 +2,17 @@
+Warning:
+    For developers:
+
+    Do not bring the ``typing.Protocol`` class into any form into the mixin framework
```
Done.

vlipovac commented on src/porepy/models/solution_strategy.py:
```diff
@@ -244,32 +262,48 @@ def reset_state_from_file(self) -> None:
+        # NOTE this will change with the generalization of the solid
+        self.solid = solid
+
+        # Store the default fluid constants to be accessible by in
```
Improved comment and removed this ominous usage of `_default_fluid_constants` (see https://github.com/pmgbergen/porepy/pull/1252#discussion_r1823405131 )

vlipovac commented on tests/compositional/test_fluid_mixtures.py:
```diff
@@ -133,35 +135,26 @@ def test_mixture_contexts(
-        "omega": 1.0,
-    }
-
     # h2o = composit.Component.from_species(composit.load_species(["H2O"])[0])
```
Cleaned.

keileg replied:
Can the commented lines, like L138, 144-46 be deleted?

keileg replied:
Sorry, vscode confused me. Resolving.

vlipovac commented on tests/models/test_fluid_mass_balance.py:
```diff
@@ -374,8 +378,15 @@ def test_ad_operator_methods_single_phase_flow(
+    for name in method_namespace:
+        method = getattr(owner, name)
+        owner = method
+    # method: Callable = getattr(model_setup, method_name)
```
Cleaned.

vlipovac commented on src/porepy/compositional/compositional_mixins.py:
```diff
@@ -26,21 +26,43 @@
+    dependencies: Sequence[Callable[[pp.GridLikeSequence], pp.ad.Variable]],
+) -> pp.ad.SurrogateFactory:
+    """Helper function to get a surrogate factory as a phase property, by providing
+    the name, the mD grid and the list of dependencies (like pressure, temperature)."""
```
Done.

keileg commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,372 @@
+ExtendedDomainFunctionType = pp.ExtendedDomainFunctionType
+
+
+class FluidDensityFromPressure(PorePyModel):
```
This is the class that should be mixed in (before the FluidMixin) to get the standard mathematician's fluid? I suggest to expand the docstring with this information.

keileg replied:
Similarly, if I want to implement a new constitutive law (say, \rho=exp(cT*p)), I can use the current implementations as a template?

vlipovac replied:
Correct. I improved the module docstring and provided a simple example highlighting that.

And yes, the classes here are great examples for how to provide the callables for phase properties.

keileg replied:
Deleted comment on SurrogateFactory (which risked scaring off users), added a comment about how to implement other simple constitutive laws.

keileg commented on src/porepy/models/constitutive_laws.py:
```diff
@@ -11,6 +11,12 @@
 import porepy as pp
 from porepy.models.protocol import PorePyModel
 
+from .fluid_property_library import *  # noqa: F403, F401
```
Actually yes (!) in this case. In effect, this is just a splitting of the implementation into different parts. Moreover, the use of __all__ in the fluid property library makes the import much more palatable.

IvarStefansson commented on src/porepy/examples/flow_benchmark_2d_case_1.py:
```diff
@@ -12,43 +12,36 @@
 
 import porepy as pp
 from porepy.applications.discretizations.flux_discretization import FluxDiscretization
+from porepy.compositional.materials import SolidConstants, _HashableDict
```
Does the import suggest that the class shouldn't be private?

keileg replied:
If so, the location in compositional.materials is also strange. 

If I understand correctly, the import is needed to allow typing of extensions of the material constants? If yes, I believe this is something we want to allow, so should we find a better home for it?

vlipovac replied:
I am open for any change in location for better fit.

vlipovac replied:
With the latest changes, there is no need for a hashable dict and I removed that class.
Can we resolve this?

IvarStefansson commented on src/porepy/models/fluid_property_library.py:
```diff
@@ -0,0 +1,396 @@
+
+    """
+
+    def fluid_specific_heat_capacity(self, subdomains: list[pp.Grid]) -> pp.ad.Operator:
```
Would it not make sense to allow boundaries?

vlipovac replied:
It would. Should we bother, despite this being called only inside the scope of the enthalpy function?

keileg replied:
I can see this being useful for setting BCs. I suggest we make an issue that takes note of this, together with other unresolved comments.

vlipovac replied:
Created #1259

vlipovac commented:
I reworked the concept of constant reference pressure and temperature values:

- The new class is found [here](https://github.com/pmgbergen/porepy/blob/066144c87e6c47ee745b1ac4a8fad4603209a191/src/porepy/compositional/materials.py#L390). This is a preliminary place, we can also move it to `models/abstract_equations.py` where the VariableMixin is found.
- Sol.Strat. is adapted to have a new [attribute ](https://github.com/pmgbergen/porepy/blob/066144c87e6c47ee745b1ac4a8fad4603209a191/src/porepy/models/solution_strategy.py#L71), which represents the reference values. It is in principle analogous to the current solid constants.
- [perturbation_from_reference ](https://github.com/pmgbergen/porepy/blob/066144c87e6c47ee745b1ac4a8fad4603209a191/src/porepy/models/abstract_equations.py#L122)was modified to wrap the reference values into AD and return `A-A_ref`
- extended material constants for testing were split, having now a [dedicated part for reference variable values](https://github.com/pmgbergen/porepy/blob/fluid_concept_2/src/porepy/applications/material_values/reference_values.py)
- An example usage is given in the tests, [like here](https://github.com/pmgbergen/porepy/blob/066144c87e6c47ee745b1ac4a8fad4603209a191/tests/models/test_momentum_balance.py#L129), and multiple other (all in `tests/models` and `tests/functional`)
- `VariableMixin.reference_*` is removed, and the abomination of `model.fluid.reference_component.pressure` should be replaced everywhere by `model.reference_variable_values.pressure`.
- [This here](https://github.com/pmgbergen/porepy/blob/066144c87e6c47ee745b1ac4a8fad4603209a191/src/porepy/models/constitutive_laws.py#L2909) was the only place which did not use perturbation_from_reference, I fixed that.
- all other changes should be namespace changes due to above.

I think a test dedicated to the reference values alone, the evaluation of the perturbation from reference and so on, is called for.

keileg commented:
@vlipovac: Regarding the tutorials, could you run these on a system with pypardiso installed, so that we don't introduce more error messages in the notebook output?

vlipovac commented:
> @vlipovac: Regarding the tutorials, could you run these on a system with pypardiso installed, so that we don't introduce more error messages in the notebook output?

Done. There is one last error after the last Python cell in `stress_discretization,ipynb` coming from a `matplotlib.Axes` object, which I am not sure how to remove.

vlipovac commented on src/porepy/applications/boundary_conditions/model_boundary_conditions.py:
```diff
@@ -118,7 +115,9 @@ def bc_values_pressure(self, boundary_grid: pp.BoundaryGrid) -> np.ndarray:
         values = np.zeros(boundary_grid.num_cells)
-        values[domain_sides.north + domain_sides.south] = self.fluid.pressure()
+        values[domain_sides.north + domain_sides.south] = (
+            self.fluid.reference_component.pressure
```
Can we resolve this? Should be solved with the introduction of `ReferenceVariableValues`

vlipovac commented:
The latest changes involve:

- renaming `MaterialConstants` to `Constants`
- renaming `FluidConstants` to `FluidComponent` and adding `compositional.base.Component` to its inheritance tree
- implementing `__setattr__` to imitate the `frozen` functionality of `dataclass`, but limited to the constants fields declares with SI units (also, inheriting from Component the `Component.fraction` must not be frozen)
- Adding generic typing to `Fluid`, `Phase`, `Compound` such that custom material parameters are found by linters and static type checkers:

   - `Compound[FluidComponent]` says the tracers in the compound are of type `FluidComponent`
   - `Phase[FluidComponent]`: says this is a phase composed of `FluidComponent`
   - `Fluid[FluidComponent, Phase[FluidComponent]]`: says this is a fluid mixture of fluid-like component and respectively phases.

Minor open todos:
1. Decide on whether to leave `compositional.materials.Constants` and `compositional.materials.ReferenceVariableValues` there or find a better spot (maybe `porepy.utils`?)
2. Add parametrized test for children of `Constants`, check for freezing of constants and proper unit conversion (though the conversion is to some degree tested separately in each model)
3. Add test for parsing and evaluation of `perturbation_from_reference` which now accesses the reference variable values.

keileg commented:
Thanks @vlipovac. I'll have a look now.

keileg commented:
@vlipovac The changes look good to me, I have no further comments to that part.

As for the remaining todos:
1. I prefer not to put things into pp.utils, so for my part, the classes can stay unless someone comes up with a much better suggestion.
2. Yes, some simple testing would be useful here.
3. Yes please.

There are three unresolved conversations that all seem to point to the creation of one (or more?) issue. I suggest we make this and then resolve.

IvarStefansson commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,534 @@
+
+
+@dataclass(kw_only=True)
+class SolidConstants(Constants):
```
Should this be SolidComponent for consistency?

keileg replied:
I would say no, or not yet. The class does not inherit from component and has no fraction.

IvarStefansson commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,534 @@
+of :class:`Constants` correctly."""
+
+
+@dataclass(kw_only=True, eq=False)
```
Why is eq set to false here? And is there a reason why it's true by default in SolidConstants?

IvarStefansson commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,534 @@
+of :class:`Constants` correctly."""
+
+
+@dataclass(kw_only=True, eq=False)
```
Why is eq set to false here? And is there a reason why it's true by default in SolidConstants?

vlipovac replied:
No need currently. The `eq=False` for `FluidComponent` is in order to make it hashable, due to omitting `frozen=True`.
Hashability must be given because `FluidComponent` is used as a dict key at various points.

`SolidConstants` is not there yet, so to speak.
But I can add it if wished! There is no point in creating an equality overload based on field values, since the dataclass is meant for representing distinct solids.

keileg replied:
For future reference, the issues with hashability for (eq=True,frozen=False) is just how dataclasses works, see https://stackoverflow.com/a/52390734.

I made the SolidConstants eq=False, and wrote a short discussion of why this happens. Could you have a look, @vlipovac, correct/improve if you think needed, and then resolve?

IvarStefansson commented on src/porepy/applications/material_values/fluid_values.py:
```diff
@@ -4,7 +4,7 @@
 
 The dictionary containing parameter values is obtained by e.g. pp.fluid_values.water.
-They can be used in a simulation by passing `fluid = pp.FluidConstants(pp.fluid_values.
+They can be used in a simulation by passing `fluid = pp.FluidComponent(pp.fluid_values.
```
```suggestion
They can be used in a simulation by passing `fluid = pp.FluidComponent(**pp.fluid_values.
```

IvarStefansson commented on src/porepy/compositional/materials.py:
```diff
@@ -0,0 +1,543 @@
+# errors are raised (no separate checks required).
+# 2. By providing default values to the fields, not every constant is required. The user
+# is expected to be aware which physics are used in the model.
+# 3. By instructing dataclass to not override the __eq__ provided by object (we do not
```
Please fix sentence. The by suggests a "By [action], [something results]" structure.

vlipovac replied:
Done.

vlipovac commented:
I think we are done.

1. Added test for perturbation from reference method in `tests/models/test_constitutive_laws/test_perturbation_from_reference`
2. Added parametrized test for all Constants classes available so far in `tests/compositional/test_constants_and_units` (formerly called `test_material_constants`)

`Constants` and `ReferenceVariableValues` are left for now in `compositional/materials`.
But in future we might want to put these two somewhere else (also having #1249 in mind)

keileg reviewed (approved):
Thanks for the effort!
